### DISCUSSION TITLE
Reducing boilerplate code using (semi) auto-implemented interfaces

### DISCUSSION THREAD

**Discussion by MrMatthewLayton** [dotnet/csharplang#7032] [Language Ideas]
When writing C#, I tend to write a lot of value type objects that implement `IEquatable<T>`, `IComparable<T>`, or both. 

For the sake of this proposal, let's assume that I'm writing a fictitious struct called `Int256` with equatable and comparable value semantics; for example:

```csharp
public readonly struct Int256 : IEquatable<Int256>, IComparable<Int256>, IComparable
{
    public bool Equals(Int256 other)
    {
        // TODO : is this equal to other?
    }

    public int CompareTo(Int256 other)
    {
        // TODO : how does this compare to other?
    }

    public int CompareTo(object? obj)
    {
        if (obj is null) return 1;
        if (obj is not Int256 int256) throw new ArgumentException("Obj must be of type Int256.");
        return CompareTo(int256);
    }

    public static bool operator ==(Int256 left, Int256 right)
    {
        return Equals(left, right);
    }
    
    public static bool operator !=(Int256 left, Int256 right)
    {
        return !Equals(left, right);
    }
    
    public static bool operator >(Int256 left, Int256 right)
    {
        return left.CompareTo(right) is 1;
    }
    
    public static bool operator >=(Int256 left, Int256 right)
    {
        return left.CompareTo(right) is 1 or 0;
    }
    
    public static bool operator <(Int256 left, Int256 right)
    {
        return left.CompareTo(right) is -1;
    }
    
    public static bool operator <=(Int256 left, Int256 right)
    {
        return left.CompareTo(right) is -1 or 0;
    }
}
```

Let's also assume that I were to create some other fictitious structs with the same semantics; for example `UInt256` and `Decimal256`. While trivial, those operators become tedious to implement for every value type object.

Recently, I've been looking at C# 11's new language features, specifically static interface methods, which I believe is largely what makes the new generic math interfaces possible. With that in mind, I could add some additional interfaces to my implementation, specifically `IEqualityOperators<TSelf, TOther, TResult>` and `IComparisonOperators<TSelf, TOther, TResult>`; for example:

```csharp
public readonly struct Int256 : IEquatable<Int256>, IComparable<Int256>, IComparable, IComparisonOperators<Int256, Int256, bool>
{
  ...
}
```

_For completeness, there is no need to implement `IEqualityOperators<TSelf, TOther, TResult>` here, as they are extended by `IComparisonOperators<TSelf, TOther, TResult>` anyway._

Ultimately, this doesn't really solve the problem. All these interfaces do is ensure that the operators are implemented, but it doesn't provide an implementation for boilerplate code.

My proposal is whether interfaces could be designed that _auto-implement_ **some** of the typical boilerplate code associated with `IEquatable<T>` and `IComparable<T>`, specifically the operators: `==`, `!=`, `>`, `>=`, `<`, `<=`; for example:

**IAutoEquatable&lt;T&gt;**

```csharp
public interface IAutoEquatable<T> : IEquatable<T>, IEqualityOperators<T, T, bool> where T : IAutoEquatable<T>
{
    // Auto-implemented boilerplate.
    static virtual bool operator ==(T? left, T? right)
    {
        return Equals(left, right);
    }
    
    // Auto-implemented boilerplate.
    static virtual bool operator !=(T? left, T? right)
    {
        return !Equals(left, right);
    }
}
```

**IAutoComparable&lt;T&gt;**

```csharp
public interface IAutoComparable<T> : IComparable<T>, IComparable, IComparisonOperators<T, T, bool> where T : IAutoComparable<T>
{
    // Auto-implemented boilerplate.
    static virtual bool operator ==(T? left, T? right)
    {
        return Equals(left, right);
    }

    // Auto-implemented boilerplate.
    static virtual bool operator !=(T? left, T? right)
    {
        return !Equals(left, right);
    }

    // Auto-implemented boilerplate.
    static virtual bool operator >(T left, T right)
    {
        return left.CompareTo(right) is 1;
    }

    // Auto-implemented boilerplate.
    static virtual bool operator >=(T left, T right)
    {
        return left.CompareTo(right) is 1 or 0;
    }

    // Auto-implemented boilerplate.
    static virtual bool operator <(T left, T right)
    {
        return left.CompareTo(right) is -1;
    }

    // Auto-implemented boilerplate.
    static virtual bool operator <=(T left, T right)
    {
        return left.CompareTo(right) is -1 or 0;
    }
}
```

The intention here is that the implementor would only require implementation of `bool Equals(T other)` and `int CompareTo(T other)` respectively, but, given that the operators are implemented on the interface, they get the operators for free!

Given my `Int256` example, it might look something like this:

```csharp
public readonly struct Int256 : IEquatable<Int256>, IComparable<Int256>, IComparable
{
    public bool Equals(Int256 other)
    {
        // TODO : is this equal to other?
    }

    public int CompareTo(Int256 other)
    {
        // TODO : how does this compare to other?
    }

    public int CompareTo(object? obj)
    {
        if (obj is null) return 1;
        if (obj is not Int256 int256) throw new ArgumentException("Obj must be of type Int256.");
        return CompareTo(int256);
    }
}
```

But I would still be able to use the operators with it; for example:

```csharp
Int256 a = 123;
Int256 b = 456;

a == b; // False
a != b; // True
a > b;  // False
a >= b; // False
a < b; 	// True
a <= b; // True
```

There is a problem, however.

Whilst those interfaces `IAutoEquatable<T>` and `IAutoComparable<T>` contain implementations for the operators, I'm still expected to implement them in `Int256`.

**Questions**

1. Why do `virtual` default implementations in interfaces still require implementation? i.e. why doesn't `Int256` just use the default implementation?
2. Might it be possible for a future version of C# to address this issue, such that we can use it to alleviate the need to write boilerplate code?

**Additional Thoughts**

One question that's been asked a few times with regards to auto-implementing `IEquatable<T>` specifically is:

> Don't records do this?

Yes, records provide an implementation of `IEquatable<T>` and override the default `Equals` and `HashCode` methods in order to provide value semantics.

**This isn't the same thing, which becomes even more evident when you consider `IComparable<T>`.**

Whilst records are great for providing structural equality and immutability, they aren't always the right design choice. Consider that there are some unusual gotchas when dealing with structural equality of records; for example:

```csharp
sealed record DoubleWrapper(double Value);

static void Main(string[] args)
{
    const double value1 = double.NaN;
    const double value2 = double.NaN;

    DoubleWrapper record1 = new(value1);
    DoubleWrapper record2 = new(value2);
    
    Console.WriteLine(value1 == value2); // False
    Console.WriteLine(record1 == record2); // True
}
```

Given that **NaN != NaN**, you might expect that records wrapping **NaN** would also return false when compared, but they don't. In this case, you would need to override `Equals` anyway; for example:

```csharp
sealed record DoubleWrapper(double Value)
{
    public bool Equals(DoubleWrapper? other)
    {
        return ReferenceEquals(this, other)
               || other is not null
               && !double.IsNaN(Value)
               && !double.IsNaN(other.Value)
               && other.Value == Value;
    }
}

static void Main(string[] args)
{
    const double value1 = double.NaN;
    const double value2 = double.NaN;

    DoubleWrapper record1 = new(value1);
    DoubleWrapper record2 = new(value2);
    
    Console.WriteLine(value1 == value2); // False
    Console.WriteLine(record1 == record2); // False
}
```

Similarly, records that contain collections return false, even when the sequences are equal; for example:

```csharp
sealed record Party(List<string> Names);

private static void Main(string[] args)
{
    Party party1 = new(new List<string> { "Alice", "Bob", "Charlie" });
    Party party2 = new(new List<string> { "Alice", "Bob", "Charlie" });
    
    Console.WriteLine(party1 == party2); // False
}
```

And so once again, `Equals` needs to be overridden in order to provide equality for collections; for example:

```csharp
sealed record Party(List<string> Names)
{
    public bool Equals(Party? other)
    {
        return ReferenceEquals(this, other)
               || other is not null
               && other.Names.SequenceEqual(Names);
    }
}

private static void Main(string[] args)
{
    Party party1 = new(new List<string> { "Alice", "Bob", "Charlie" });
    Party party2 = new(new List<string> { "Alice", "Bob", "Charlie" });
    
    Console.WriteLine(party1 == party2) // True;
}
```

So, whilst records can be great for structural equality, they certainly don't fit all cases.

But, I think records could benefit from this proposal too, given that records implement the `==` and `!=` operators, record implementations in future could implement `IAutoEquatable<T>`, instead of `IEquatable<T>`, where the compiler would not need to implement the operators, because the implementtion could be obtained from the interface.

**References**

Question raised on [StackOverflow](https://stackoverflow.com/questions/75643829/is-it-possible-to-auto-implement-iequatablet-and-icomparablet-in-c-sharp-11)

**Comment by HaloFour** [comment]
> 1. Why do `virtual` default implementations in interfaces still require implementation? i.e. why doesn't `Int256` just use the default implementation?

Interfaces are not inherited, the default implementation exists but only if you call through the interface.  Those members don't exist on `Int256`.  You'd typically use these features with constrained generics.

> 2. Might it be possible for a future version of C# to address this issue, such that we can use it to alleviate the need to write boilerplate code?

Today you could use an analyzer with a fixer to automatically generate all of that boilerplate code for you.  Or a source generator, such as the following:

https://github.com/diegofrata/Generator.Equals
